test/CMakeLists.txt

```diff
@@ -81,17 +81,23 @@ if(WITH_FUZZERS)
     add_subdirectory(fuzz)
 endif()
 
-if(CMAKE_VERSION VERSION_LESS 3.12)
+if((WITH_GTEST OR WITH_BENCHMARKS) AND CMAKE_VERSION VERSION_LESS 3.12)
     message(WARNING "Minimum cmake version of 3.12 not met for GoogleTest or benchmarks!")
-else()
+    set(WITH_BENCHMARKS OFF PARENT_SCOPE)
+    set(WITH_GTEST OFF PARENT_SCOPE)
+    set(WITH_BENCHMARKS OFF)
+    set(WITH_GTEST OFF)
+endif()
+
+if(WITH_BENCHMARKS)
+    add_subdirectory(benchmarks)
+endif()
```

Comment:
This is technically incorrect, as the benchmarks are written in C++, but the language isn't enabled yet at this point, so if `WITH_GTEST` is disabled, the compile might fail or the result is unpredictable.

Reply:
It's enabled for the benchmarks in the subdirectory's CMakeLists.txt file: https://github.com/zlib-ng/zlib-ng/blob/1a96258117775492378a4e62d070ade203d2aa77/test/benchmarks/CMakeLists.txt#L8

Reply:
The cmake manual says that it has to be enabled in earliest possible file, not where it is actually first used. This is because the result of the command is inherited by child files, but not propagated back to parent file, when the child file is finished processing. This might cause issues with caching the results, thus the result becoming unpredictable.

Reply:
The only children of the main CMakeLists.txt are the tests directory which is the only place it uses C++. I think it is fine.

Reply:
Actually `enable_language(CXX)` should be moved up in this file, as part of `(WITH_GTEST OR WITH_BENCHMARKS)` check.

Reply:
Maybe something like this?
```
if(WITH_GTEST OR WITH_BENCHMARKS)
    if(CMAKE_VERSION VERSION_LESS 3.12)
        message(WARNING "Minimum cmake version of 3.12 not met for GoogleTest or benchmarks!")
        set(WITH_BENCHMARKS OFF PARENT_SCOPE)
        set(WITH_GTEST OFF PARENT_SCOPE)
        set(WITH_BENCHMARKS OFF)
        set(WITH_GTEST OFF)
    else()
        enable_language(CXX)
    endif()
endif()
```

Reply:
@nmoinvaz That is/was pretty much what I thought would be the best solution...

Reply:
@ccawley2011 can you make this change?

Reply:
OK, done.